// @crayonai/react-ui/dist/components/Charts/BarChartCondensed/BarChartCondensed.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import React, { useCallback, useEffect, useMemo, useRef, useState } from "react";
import { Bar, BarChart as RechartsBarChart, XAxis, YAxis } from "recharts";
import { useId } from "../../../polyfills";
import { useTheme } from "../../ThemeProvider";
import { ChartContainer, ChartTooltip } from "../Charts";
import { DEFAULT_X_AXIS_HEIGHT, X_AXIS_PADDING } from "../constants";
import { SideBarTooltipProvider } from "../context/SideBarTooltipContext";
import { useAutoAngleCalculation, useMaxLabelWidth, useTransformedKeys, useYAxisLabelWidth, } from "../hooks";
import { cartesianGrid, CustomTooltipContent, DefaultLegend, LineInBarShape, SVGXAxisTick, YAxisTick, } from "../shared";
import { LabelTooltipProvider } from "../shared/LabelTooltip/LabelTooltip";
import { getBarStackInfo, getRadiusArray } from "../utils/BarCharts/BarChartsUtils";
import { get2dChartConfig, getDataKeys, getLegendItems } from "../utils/dataUtils";
import { useChartPalette } from "../utils/PalletUtils";
const BAR_WIDTH = 12;
const BAR_GAP = 10;
const BAR_CATEGORY_GAP = "20%";
const BAR_INTERNAL_LINE_WIDTH = 1;
const BAR_RADIUS = 4;
const CHART_HEIGHT = 296;
const CHART_CONTAINER_BOTTOM_MARGIN = 10;
const BarChartCondensedComponent = ({ data, categoryKey, theme = "ocean", customPalette, variant = "grouped", tickVariant = "singleLine", grid = true, icons = {}, radius, isAnimationActive = false, showYAxis = true, xAxisLabel, yAxisLabel, legend = true, className, height = CHART_HEIGHT, width, }) => {
    const dataKeys = useMemo(() => {
        return getDataKeys(data, categoryKey);
    }, [data, categoryKey]);
    const { yAxisWidth, setLabelWidth } = useYAxisLabelWidth(data, dataKeys);
    const maxLabelWidth = useMaxLabelWidth(data, categoryKey);
    const chartContainerRef = useRef(null);
    const [chartContainerWidth, setChartContainerWidth] = useState(0);
    const widthOfData = useMemo(() => {
        if (data.length === 0) {
            return 0;
        }
        return chartContainerWidth / data.length;
    }, [chartContainerWidth, data]);
    const { angle: calculatedAngle, height: xAxisHeight } = useAutoAngleCalculation(maxLabelWidth, tickVariant === "angled", maxLabelWidth < 100 ? widthOfData : undefined);
    const isAngled = useMemo(() => {
        return calculatedAngle !== 0;
    }, [calculatedAngle]);
    const effectiveHeight = useMemo(() => {
        if (tickVariant === "angled") {
            return xAxisHeight + height;
        }
        return height + DEFAULT_X_AXIS_HEIGHT;
    }, [height, xAxisHeight, tickVariant]);
    const transformedKeys = useTransformedKeys(dataKeys);
    const colors = useChartPalette({
        chartThemeName: theme,
        customPalette,
        themePaletteName: "barChartPalette",
        dataLength: dataKeys.length,
    });
    const chartConfig = useMemo(() => {
        return get2dChartConfig(dataKeys, colors, transformedKeys, undefined, icons);
    }, [dataKeys, icons, colors, transformedKeys]);
    const id = useId();
    const chartMargin = useMemo(() => ({
        top: 10,
        right: 10,
        bottom: CHART_CONTAINER_BOTTOM_MARGIN,
        left: showYAxis ? 10 : 0,
    }), [showYAxis]);
    const { mode, theme: userTheme } = useTheme();
    const calculatedRadius = useMemo(() => {
        let radiusValue = BAR_RADIUS;
        if (typeof radius === "number") {
            radiusValue = radius;
        }
        else {
            const radiusTheme = userTheme.rounded2xs;
            if (radiusTheme) {
                radiusValue = typeof radiusTheme === "string" ? parseInt(radiusTheme) : radiusTheme;
            }
        }
        return radiusValue;
    }, [userTheme.rounded2xs, radius]);
    const barInternalLineColor = useMemo(() => {
        if (mode === "light") {
            return "rgba(255, 255, 255, 0.3)";
        }
        return "rgba(0, 0, 0, 0.3)";
    }, [mode]);
    const containerRef = useRef(null);
    const [containerWidth, setContainerWidth] = useState(0);
    const [isSideBarTooltipOpen, setIsSideBarTooltipOpen] = useState(false);
    const [sideBarTooltipData, setSideBarTooltipData] = useState({
        title: "",
        values: [],
    });
    const [hoveredCategory, setHoveredCategory] = useState(null);
    const [isLegendExpanded, setIsLegendExpanded] = useState(false);
    // Use provided width or observed width
    const effectiveWidth = useMemo(() => {
        return width ?? containerWidth;
    }, [width, containerWidth]);
    // Handle mouse events for bar hovering
    const handleChartMouseMove = useCallback((state) => {
        if (state && state.activeLabel !== undefined) {
            setHoveredCategory(state.activeLabel);
        }
    }, []);
    const handleChartMouseLeave = useCallback(() => {
        setHoveredCategory(null);
    }, []);
    // Observe container width for legend
    useEffect(() => {
        // Only set up ResizeObserver if width is not provided
        if (width || !containerRef.current || !chartContainerRef.current) {
            return () => { };
        }
        const resizeObserver = new ResizeObserver((entries) => {
            // there is only one entry in the entries array because we are observing the chart container
            for (const entry of entries) {
                if (entry.target === containerRef.current) {
                    setContainerWidth(entry.contentRect.width);
                }
                if (entry.target === chartContainerRef.current) {
                    setChartContainerWidth(entry.contentRect.width);
                }
            }
        });
        resizeObserver.observe(containerRef.current);
        resizeObserver.observe(chartContainerRef.current);
        return () => {
            resizeObserver.disconnect();
        };
    }, [width]);
    useEffect(() => {
        setIsLegendExpanded(false);
    }, [dataKeys]);
    // Memoize legend items creation
    const legendItems = useMemo(() => {
        if (!legend) {
            return [];
        }
        return getLegendItems(dataKeys, colors, icons);
    }, [dataKeys, colors, icons, legend]);
    const yAxis = useMemo(() => {
        if (!showYAxis) {
            return null;
        }
        return (_jsx("div", { className: "crayon-bar-chart-condensed-y-axis-container", children: _jsxs(RechartsBarChart, { width: yAxisWidth, height: effectiveHeight, data: data, stackOffset: "sign", margin: {
                    top: chartMargin.top,
                    bottom: xAxisHeight + chartMargin.bottom, // this is required to give space for x-axis
                    left: 0,
                    right: 0,
                }, children: [_jsx(YAxis, { width: yAxisWidth, tickLine: false, axisLine: false, tick: _jsx(YAxisTick, { setLabelWidth: setLabelWidth }) }), dataKeys.map((key) => {
                        return (_jsx(Bar, { dataKey: key, fill: "transparent", stackId: variant === "stacked" ? "a" : undefined, isAnimationActive: false, maxBarSize: 0 }, `yaxis-bar-chart-condensed-${key}`));
                    })] }, `y-axis-bar-chart-condensed-${id}`) }));
    }, [
        showYAxis,
        effectiveHeight,
        data,
        dataKeys,
        variant,
        id,
        yAxisWidth,
        chartMargin,
        xAxisHeight,
        setLabelWidth,
    ]);
    const barElements = useMemo(() => {
        return dataKeys.map((key) => {
            const transformedKey = transformedKeys[key];
            const color = `var(--color-${transformedKey})`;
            return (_jsx(Bar, { dataKey: key, fill: color, stackId: variant === "stacked" ? "a" : undefined, isAnimationActive: isAnimationActive, maxBarSize: BAR_WIDTH, barSize: BAR_WIDTH, shape: (props) => {
                    const { payload, value, dataKey } = props;
                    const { isNegative, isFirstInStack, isLastInStack, hasNegativeValueInStack } = getBarStackInfo(variant, value, dataKey, payload, dataKeys);
                    const customRadius = getRadiusArray(variant, calculatedRadius, "vertical", isFirstInStack, isLastInStack, isNegative);
                    return (_jsx(LineInBarShape, { ...props, radius: customRadius, internalLineColor: barInternalLineColor, internalLineWidth: BAR_INTERNAL_LINE_WIDTH, isHovered: hoveredCategory !== null, hoveredCategory: hoveredCategory, categoryKey: categoryKey, variant: variant, hasNegativeValueInStack: hasNegativeValueInStack }));
                } }, `bar-${key}`));
        });
    }, [
        dataKeys,
        transformedKeys,
        variant,
        calculatedRadius,
        isAnimationActive,
        barInternalLineColor,
        hoveredCategory,
        categoryKey,
    ]);
    return (_jsx(LabelTooltipProvider, { children: _jsx(SideBarTooltipProvider, { isSideBarTooltipOpen: isSideBarTooltipOpen, setIsSideBarTooltipOpen: setIsSideBarTooltipOpen, data: sideBarTooltipData, setData: setSideBarTooltipData, children: _jsxs("div", { className: clsx("crayon-bar-chart-condensed-container", className), style: {
                    width: width ? `${width}px` : undefined,
                }, children: [yAxisLabel && (_jsx("div", { className: "crayon-bar-chart-condensed-y-axis-label", children: yAxisLabel })), _jsxs("div", { className: "crayon-bar-chart-condensed-container-inner", ref: containerRef, children: [yAxis, _jsx("div", { className: "crayon-bar-chart-condensed", ref: chartContainerRef, children: _jsx(ChartContainer, { config: chartConfig, style: { width: "100%", height: effectiveHeight }, rechartsProps: {
                                        width: "100%",
                                        height: "100%",
                                    }, children: _jsxs(RechartsBarChart, { stackOffset: "sign", accessibilityLayer: true, data: data, margin: chartMargin, barGap: BAR_GAP, barCategoryGap: BAR_CATEGORY_GAP, onMouseMove: handleChartMouseMove, onMouseLeave: handleChartMouseLeave, children: [grid && cartesianGrid(), _jsx(XAxis, { dataKey: categoryKey, tickLine: false, axisLine: false, textAnchor: isAngled ? "end" : "middle", interval: "preserveStartEnd", minTickGap: 5, height: xAxisHeight, tick: _jsx(SVGXAxisTick, {}), angle: calculatedAngle, orientation: "bottom", padding: {
                                                    left: X_AXIS_PADDING,
                                                    right: X_AXIS_PADDING,
                                                } }), _jsx(ChartTooltip, { cursor: {
                                                    fill: "var(--crayon-sunk-fills)",
                                                    stroke: "var(--crayon-stroke-default)",
                                                    opacity: 1,
                                                    strokeWidth: 1,
                                                }, content: _jsx(CustomTooltipContent, { parentRef: containerRef }), offset: 10 }), barElements] }, `bar-chart-condensed-${id}`) }) })] }), xAxisLabel && (_jsx("div", { className: "crayon-bar-chart-condensed-x-axis-label", children: xAxisLabel })), legend && (_jsx(DefaultLegend, { items: legendItems, containerWidth: effectiveWidth, isExpanded: isLegendExpanded, setIsExpanded: setIsLegendExpanded }))] }) }) }));
};
// Added React.memo for performance optimization to avoid unnecessary re-renders
export const BarChartCondensed = React.memo(BarChartCondensedComponent);
